ls1intum/Artemis
# `Exam mode`: Improve performance of exam live statistics

### Checklist
#### General


#### Server
#### Client

### Motivation and Context



This PR is for testing and improving the performance of the exam live statistics dashboard with many actions (e.g. 1.000.000+ actions). In addition, i performed multiple benchmarks and verify that my changes improve the performance.

In addition, this PR fixes a small bug that caused duplicate values when switching through subsections.

https://confluence.ase.in.tum.de/display/ArTEMiS/Performance

### Description


- Reduced the amount of observer calls
- Replaced set with list
- Improved the perpetration and validation of actions

### Steps for Testing

Prerequisites:
- 1 Instructor
- Multiple Students
- 1 Exam with Exam Live Statistics enabled

1. Log in to Artemis
2. Create an Exam with Exam Live Statistics enabled
3. Register multiple Students
4. Participate with those students
5. View the Exam Live Statistics as instructor during the Exam
6. Reload the page (should work up to 1+ Million actions)
7. Switch sections (should work without any delay)
8. Verify that the provided information are correct and the old actions are also loaded

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| average-actions-chart.component.ts | 86.36% | 95.65% |
| category-actions-chart.component.ts | 83.33% | 96.77% |
| total-actions-chart.component.ts | 79.16% | 93.1% |
| chart.component.ts | 83.33% | 96% |
| exercise-chart.component.ts | 83.33 | 95% |
| exercise-group-chart.component.ts | 85.71% | 96.15% |
| exercise-navigation-chart.component.ts | 88.23% | 96% |
| exercise-submission-chart.component.ts | 88.23% | 96% |
| exam-action.service.ts | 78.43% | 98.75% |
| exam-monitoring.component.ts | 86.36% | 97.36% |
| monitoring-overview.component.ts | 75% | 100% |
